.github/workflows/check-formatting.yml

```
@@ -0,0 +1,21 @@
+name: Check formatting
+on: workflow_call
+
+jobs:
+  pre-commit:
+    name: Run pre-commit hooks
+    runs-on: ubuntu-latest
```

```suggestion
    runs-on: ubuntu-22.04
```
because format_header currently still use the system clang-format-14, using 22.04 might be more stable

---

I will change this, but what is the purpose of `format_header.sh` anyway? It seems to reorder the includes, but that should be covered by clang-format already. Special handling like the `force-on-top` could be handled by disabling clang-format.

---

no, clang-format reordering does not fit our main header (it's somehow complex, so I would also like to change) and the ordering can not be handled across sections (or across-section reorder will only contain one space line for sep?)

---

the format_header did two things:
1. put the correct main header on the top
2. reorder the header with clang-format 12+ and then replace one line sep -> two line sep.

---

There is a way to achieve 1. with clang-format. But that requires changing
```
#include <ginkgo/core/what/ever.hpp>
```
into
```
#include "ginkgo/core/what/ever.hpp"
```
So clang-format recognizes the main header of a source file only if its in `""`, but not if its in `<>`.
I will also look into 2.

---

No, the main header issue is we have several rules for main header. Please take a look at https://github.com/ginkgo-project/ginkgo/blob/develop/dev_tools/scripts/config
Some of them could be merged, but it's still complex if we do not have certain forced rules.
We try writing the general rules in https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines#main-header
From https://clang.llvm.org/docs/ClangFormatStyleOptions.html#includeismainregex, you can only have simple regex for all of them.

Thanks for looking at that, look forward to avoiding complex format_header.
I suggest it should be another pr if you find something easy to maintain.